akka-http/src/main/scala/akka/http/scaladsl/coding/Decoder.scala

```diff
@@ -5,18 +5,24 @@
 package akka.http.scaladsl.coding
 
 import akka.NotUsed
+import akka.http.javadsl.model.HttpMessage.MessageTransformations
 import akka.http.scaladsl.model._
 import akka.stream.{ FlowShape, Materializer }
 import akka.stream.stage.GraphStage
 import akka.util.ByteString
 import headers.HttpEncoding
-import akka.stream.scaladsl.{ Sink, Source, Flow }
+import akka.stream.scaladsl.{ Flow, Sink, Source }
 
 import scala.concurrent.Future
 
 trait Decoder {
   def encoding: HttpEncoding
 
+  def decodeMessage[T <: HttpMessage with MessageTransformations[T]](message: T): T#Self =
```

Comment:
Shouldn't be the other way round? `decode` calling `decodeMessage`, otherwise what is the purpose of the `transformEntityDataBytes` method?